src/wurfapi/template_files/macros.rst

```diff
@@ -114,13 +112,21 @@ a non-character. Otherwise rst will fail with an error.
 {# FORMAT_PARAGRAPH #}
 
 {% macro format_paragraph(paragraph) %}
-{% for paragraph_element in paragraph %}
-{% if paragraph_element["kind"] == "text" %}
-{{ format_text(paragraph_element) -}}
-{% elif paragraph_element["kind"] == "code" %}
-{{ format_code(paragraph_element) -}}
-{% elif paragraph_element["kind"] == "list" %}
-{{ format_list(paragraph_element) -}}
+{% for element in paragraph %}
+{% if not loop.first %}
+{% set last_element = paragraph[loop.index0 - 1] %}
+{% set last_was_code_block = (last_element.kind == "code" and last_element.is_block) %}
+{% set startswith_punctuation = 'content' in element and element.content[0] in ',.!?:;' %}
+{%- if not startswith_punctuation and not last_was_code_block -%}
+{{ " " -}}
```

Comment:
Can't we make it so that spaces are added in the json paragraph element in the right places. Such that the jinja template can be more simple?

Reply:
I'm not sure if it's possible. In general I think it would be great if we could get some of all the complexity out of the templates. It's rather difficult to write and nearly impossible to read.

Reply:
Yes, in the ideal world one could just join all the `content` fields of all paragraphs and then you would have the text as it should look. Perhaps we can do some processing in Python before handing it over to the templates to fix up the text.

Reply:
I believe you can expose python functions to jinja2 templates. Maybe we should look into adding some RST helpers - or even better see if something already exists.